File: design/swift/HousePour/HPMenu.swift
// House Pour kit — HPMenu, HPKebabButton (PRODUCT.md §2.6).
//
// A kebab button and the menu it opens: a panel card at the card radius carrying the one card
// shadow. Wider than `columnMax` it is anchored under the button (flipping above it when there
// is no room below); at `columnMax` or narrower — every phone — it is a bottom sheet instead.
// Never SwiftUI's `Menu` — that paints system chrome. The presentation itself is
// animation-free; only the surface fades, per COMPONENTS.md rule 4.

import SwiftUI

/// One action row in an HPMenu.
public struct HPMenuItem: Identifiable {
    public let id = UUID()
    public let label: String
    public let action: () -> Void
    public init(_ label: String, action: @escaping () -> Void) {
        self.label = label; self.action = action
    }
}

/// A vertical three-dot button: ghost, `faint` dots drawn from tokens (ink while pressed), 40pt hit target.
public struct HPKebabButton: View {
    let label: String
    let action: () -> Void
    public init(label: String = "More", action: @escaping () -> Void) {
        self.label = label; self.action = action
    }
    public var body: some View {
        Button(action: action) {
            VStack(spacing: HPMetric.kebabDotGap) {
                ForEach(0..<HPMetric.kebabDots, id: \.self) { _ in
                    // No fill: the dots take the style's foreground so they can step up on press.
                    Circle().frame(width: HPMetric.kebabDot, height: HPMetric.kebabDot)
                }
            }
            .hpTouchTarget()
        }
        .buttonStyle(HPKebabStyle())
        .accessibilityLabel(label)
    }
}

/// The ghost press: dots step `faint` → ink over `Motion.color`, plus the kit's 1pt press translate.
private struct HPKebabStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .foregroundStyle(configuration.isPressed ? HPTokens.Colors.ink : HPTokens.Colors.faint)
            .animation(HPMotion.color, value: configuration.isPressed)
            .offset(y: configuration.isPressed ? HPTokens.Motion.pressTranslateY : 0)
            .animation(HPMotion.press, value: configuration.isPressed)
    }
}

/// The kebab button plus its menu. The button measures its own frame at tap time, so the surface
/// can hang under it without a preference round-trip.
public struct HPMenu: View {
    let label: String
    let items: [HPMenuItem]
    @State private var anchor: CGRect = .zero
    @State private var isPresented = false

    public init(label: String = "More", items: [HPMenuItem]) {
        self.label = label; self.items = items
    }

    public var body: some View {
        GeometryReader { geo in
            HPKebabButton(label: label) {
                anchor = geo.frame(in: .global)
                // The surface fades itself in; the presentation must not slide.
                var instant = Transaction()
                instant.disablesAnimations = true
                withTransaction(instant) { isPresented = true }
            }
        }
        .frame(width: HPTokens.Space.touchMin, height: HPTokens.Space.touchMin)
        .fullScreenCover(isPresented: $isPresented) {
            HPMenuSurface(anchor: anchor, items: items, isPresented: $isPresented)
                .presentationBackground(.clear)
        }
    }
}

/// The card's measured height, so the anchored menu knows whether it fits under the button.
private struct HPMenuHeightKey: PreferenceKey {
    static var defaultValue: CGFloat = 0
    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) { value = max(value, nextValue()) }
}

/// The scrim and the card. Dismisses on a tap outside or a swipe down on the sheet.
struct HPMenuSurface: View {
    let anchor: CGRect
    let items: [HPMenuItem]
    @Binding var isPresented: Bool
    @State private var shown = false
    @State private var cardHeight: CGFloat = 0

    var body: some View {
        GeometryReader { geo in
            let origin = geo.frame(in: .global).origin
            // COMPONENTS.md: the anchored/sheet split is the column width, not the size class.
            let isSheet = geo.size.width <= HPTokens.Space.columnMax
            ZStack(alignment: .topLeading) {
                HPTokens.Colors.scrim
                    .ignoresSafeArea()
                    .contentShape(Rectangle())
                    .onTapGesture { close() }
                    .accessibilityLabel("Dismiss")
                    .accessibilityAddTraits(.isButton)
                if isSheet {
                    card
                        .gesture(DragGesture(minimumDistance: HPMetric.menuDismissDrag)
                            .onEnded { if $0.translation.height > HPMetric.menuDismissDrag { close() } })
                        .frame(maxWidth: HPTokens.Space.columnMax)
                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
                        .padding(.horizontal, HPTokens.Space.columnSide)
                } else {
                    card
                        .frame(width: HPMetric.menuWidth)
                        .background(GeometryReader { inner in
                            Color.clear.preference(key: HPMenuHeightKey.self, value: inner.size.height)
                        })
                        .offset(x: anchoredX(in: geo.size.width, originX: origin.x),
                                y: anchoredY(in: geo.size.height, originY: origin.y))
                }
            }
            .onPreferenceChange(HPMenuHeightKey.self) { cardHeight = $0 }
        }
        .opacity(shown ? 1 : 0)
        .onAppear { withAnimation(HPMotion.toast) { shown = true } }
    }

    /// Trailing-aligned with the button, kept a column side inside both screen edges.
    private func anchoredX(in width: CGFloat, originX: CGFloat) -> CGFloat {
        let trailing = anchor.maxX - originX - HPMetric.menuWidth
        let maxX = width - HPMetric.menuWidth - HPTokens.Space.columnSide
        return min(max(HPTokens.Space.columnSide, trailing), max(HPTokens.Space.columnSide, maxX))
    }

    /// Under the button, a `rowGap` below it — above it instead when the card would run off the bottom.
    private func anchoredY(in height: CGFloat, originY: CGFloat) -> CGFloat {
        let below = anchor.maxY - originY + HPTokens.Space.rowGap
        guard cardHeight > 0, below + cardHeight > height else { return below }
        return max(HPTokens.Space.columnSide, anchor.minY - originY - HPTokens.Space.rowGap - cardHeight)
    }

    private var card: some View {
        HPListCard {
            ForEach(Array(items.enumerated()), id: \.element.id) { index, item in
                Button {
                    item.action()
                    close()
                } label: {
                    HPListItem(isLast: index == items.count - 1) {
                        HPBody(item.label)
                    }
                    .frame(minHeight: HPTokens.Space.touchMin)
                    .contentShape(Rectangle())
                }
                .buttonStyle(HPPressStyle())
                .accessibilityLabel(item.label)
            }
        }
    }

    /// Fades out over the toast duration, then tears the presentation down without an animation.
    private func close() {
        guard shown else { return }
        withAnimation(HPMotion.toast) { shown = false }
        Task { @MainActor in
            try? await Task.sleep(for: .seconds(HPTokens.Motion.toast))
            var instant = Transaction()
            instant.disablesAnimations = true
            withTransaction(instant) { isPresented = false }
        }
    }
}

File: design/kotlin/housepour/HPMenu.kt
package ca.lucianlabs.housepour

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.MutableTransitionState
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.gestures.detectVerticalDragGestures
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntRect
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Popup
import androidx.compose.ui.window.PopupPositionProvider
import androidx.compose.ui.window.PopupProperties

/**
 * The kebab: a vertical three-dot button. Ghost styling — no fill, no border — with the dots drawn from tokens
 * (three `faint` circles, never a glyph or an icon font), a `touchMin` hit target, and the ghost button's own
 * press behaviour: colour to ink over `Motion.color`, 1pt translate over `Motion.press`.
 */
@Composable
fun HPKebabButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    contentDescription: String = "More",
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val dotColor by animateColorAsState(
        targetValue = if (pressed) HPTokens.Colors.ink else HPTokens.Colors.faint,
        animationSpec = tween(HPTokens.Motion.COLOR_MS),
        label = "kebab",
    )
    val translate by animateDpAsState(
        targetValue = if (pressed) HPTokens.Motion.pressTranslateY else 0.dp,
        animationSpec = tween(HPTokens.Motion.PRESS_MS),
        label = "press",
    )
    // COMPONENTS.md: a `kebabDot` (4) dot, `kebabDotGap` (3) apart — an 18pt column of dots inside the 40pt target.
    val dot = HPTokens.Space.kebabDot
    val step = dot + HPTokens.Space.kebabDotGap
    Box(
        modifier = modifier
            .offset(y = translate)
            .size(HPTokens.Space.touchMin)
            .clip(RoundedCornerShape(HPTokens.Radius.pill))
            .clickable(interactionSource = interaction, indication = null, role = Role.Button, onClick = onClick)
            .semantics { this.contentDescription = contentDescription },
        contentAlignment = Alignment.Center,
    ) {
        Canvas(Modifier.size(width = dot, height = dot + step * 2)) {
            val radius = size.width / 2f
            val stepPx = step.toPx()
            for (i in 0..2) drawCircle(dotColor, radius = radius, center = Offset(radius, radius + stepPx * i))
        }
    }
}

/**
 * A menu: the `panel` card at the card radius with the one card shadow, holding one [HPMenuItem] per action.
 *
 * Call it inside the `Box` that holds its button — on a regular width the card is a `Popup` anchored under that
 * button, right edges aligned, flipping above it when there is no room below. On a compact width (narrower than
 * `columnMax` — every phone) it is a bottom sheet over the `scrim` instead.
 *
 * Dismissal: a tap outside, the system back gesture, or — on the sheet — a swipe down. Nothing animates on
 * appear beyond the `Motion.toast` fade.
 */
@Composable
fun HPMenu(
    expanded: Boolean,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    // Kept mounted through the fade out, so the card does not vanish the frame it is dismissed.
    val visible = remember { MutableTransitionState(false) }
    visible.targetState = expanded
    if (!visible.currentState && !visible.targetState) return

    val compact = LocalConfiguration.current.screenWidthDp.dp < HPTokens.Space.columnMax
    if (compact) HPMenuSheet(visible, onDismissRequest, modifier, content)
    else HPMenuAnchored(visible, onDismissRequest, modifier, content)
}

/** One action in a menu: an `HPListItem` row, body text in ink (accent while pressed), `touchMin` minimum height. */
@Composable
fun HPMenuItem(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    isLast: Boolean = false,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val color by animateColorAsState(
        targetValue = if (pressed) HPTokens.Colors.accent else HPTokens.Colors.ink,
        animationSpec = tween(HPTokens.Motion.COLOR_MS),
        label = "menuItem",
    )
    HPListItem(
        modifier = modifier
            .clickable(interactionSource = interaction, indication = null, role = Role.Button, onClick = onClick)
            .defaultMinSize(minHeight = HPTokens.Space.touchMin)
            .semantics { contentDescription = label },
        isLast = isLast,
    ) {
        HPBody(label, color = color, maxLines = 1)
    }
}

/** The menu surface itself: `panel`, card radius, 1pt `line`, the one card shadow. */
@Composable
private fun HPMenuCard(modifier: Modifier = Modifier, content: @Composable ColumnScope.() -> Unit) {
    val shape = RoundedCornerShape(HPTokens.Radius.card)
    Column(
        modifier = modifier
            .widthIn(min = HPTokens.Space.menuWidth, max = HPTokens.Space.columnMax)
            .hpShadow(HPTokens.Radius.card, HPTokens.Shadow.contact, HPTokens.Shadow.cast)
            .clip(shape)
            .background(HPTokens.Colors.panel, shape)
            .border(HPTokens.BORDER_WIDTH.dp, HPTokens.Colors.line, shape)
            .padding(horizontal = HPTokens.Space.cardPad, vertical = HPTokens.Space.rowGap),
        content = content,
    )
}

@Composable
private fun HPMenuAnchored(
    visible: MutableTransitionState<Boolean>,
    onDismissRequest: () -> Unit,
    modifier: Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    val gap = with(LocalDensity.current) { HPTokens.Space.rowGap.roundToPx() }
    val provider = remember(gap) { HPMenuPositionProvider(gap) }
    Popup(
        popupPositionProvider = provider,
        onDismissRequest = onDismissRequest,
        properties = PopupProperties(focusable = true),
    ) {
        AnimatedVisibility(
            visibleState = visible,
            enter = fadeIn(tween(HPTokens.Motion.TOAST_MS)),
            exit = fadeOut(tween(HPTokens.Motion.TOAST_MS)),
        ) {
            HPMenuCard(modifier, content)
        }
    }
}

@Composable
private fun HPMenuSheet(
    visible: MutableTransitionState<Boolean>,
    onDismissRequest: () -> Unit,
    modifier: Modifier,
    content: @Composable ColumnScope.() -> Unit,
) {
    // A swipe past `menuDismissDrag` (one hit target) dismisses; the card itself never moves (no transforms in the look).
    val threshold = with(LocalDensity.current) { HPTokens.Space.menuDismissDrag.toPx() }
    val dragged = remember { mutableFloatStateOf(0f) }
    Popup(
        popupPositionProvider = HPMenuSheetPositionProvider,
        onDismissRequest = onDismissRequest,
        properties = PopupProperties(focusable = true),
    ) {
        AnimatedVisibility(
            visibleState = visible,
            enter = fadeIn(tween(HPTokens.Motion.TOAST_MS)),
            exit = fadeOut(tween(HPTokens.Motion.TOAST_MS)),
        ) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(HPTokens.Colors.scrim)
                    .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, onClick = onDismissRequest)
                    .systemBarsPadding(),
                contentAlignment = Alignment.BottomCenter,
            ) {
                HPMenuCard(
                    modifier
                        .padding(HPTokens.Space.columnSide)
                        .fillMaxWidth()
                        .pointerInput(threshold) {
                            detectVerticalDragGestures(
                                onDragStart = { dragged.floatValue = 0f },
                                onDragEnd = { if (dragged.floatValue > threshold) onDismissRequest() },
                                onDragCancel = { dragged.floatValue = 0f },
                            ) { change, dy ->
                                change.consume()
                                dragged.floatValue += dy
                            }
                        }
                        .pointerInput(Unit) { detectTapGestures { } },
                    content,
                )
            }
        }
    }
}

/** Under the button, right edges aligned; above it when the card would fall off the bottom. */
private class HPMenuPositionProvider(private val gap: Int) : PopupPositionProvider {
    override fun calculatePosition(
        anchorBounds: IntRect,
        windowSize: IntSize,
        layoutDirection: LayoutDirection,
        popupContentSize: IntSize,
    ): IntOffset {
        val maxX = (windowSize.width - popupContentSize.width).coerceAtLeast(0)
        val x = (anchorBounds.right - popupContentSize.width).coerceIn(0, maxX)
        val below = anchorBounds.bottom + gap
        val y = if (below + popupContentSize.height <= windowSize.height) below
        else (anchorBounds.top - gap - popupContentSize.height).coerceAtLeast(0)
        return IntOffset(x, y)
    }
}

/** The sheet covers the window; the scrim inside it does the rest. */
private object HPMenuSheetPositionProvider : PopupPositionProvider {
    override fun calculatePosition(
        anchorBounds: IntRect,
        windowSize: IntSize,
        layoutDirection: LayoutDirection,
        popupContentSize: IntSize,
    ): IntOffset = IntOffset.Zero
}
